### DISCUSSION TITLE
Extend the behaviour of the media picker

### DISCUSSION THREAD

**Discussion by MarcBruins** [OrchardCMS/OrchardCore#6874] [Q&A] [answered]
For my current project we have the requirement to be able to upload a zip(from an external party) extract the zip, apply some logic to the contents and then show it(like a zip again) in the media picker.

The most elegant solution in my opinion would be to extend the behaviour of the media picker, check for the file extension and if its zip and it contains `contents.xml` or something similiar apply the custom logic and let the upload flow continue.

Is this something that can be done? Or would it be better to for example add a custom upload module where they can upload there contents?

**Comment by MarcBruins** [comment]
So i've found the [IMediaCreatingEventHandler](https://github.com/OrchardCMS/OrchardCore/blob/dev/src/OrchardCore/OrchardCore.Media.Abstractions/Events/IMediaCreatingEventHandler.cs). I'm going to experimented with that a bit and update this discussion with my new findings

**Reply by Piedone** [reply]
Do you want to actually save the newly created zip in the Media Library or just make it appear as it exists there when using the media picker?

**Reply by MarcBruins** [reply]
So the general idea is this, we export a self content package from a LMS system, this is a html, css and some config. This export is a .zip, now from a .zip we cannot host anything. We can however host it if we extract the zip and store it on for example azure blob storage(azure.media module). 

So it should exist in the media picker(because that is in our case a gui for a blob storage) but it should exist as a folder or maybe a custom extension and not as a zip and we should be able to select it

**Reply by Piedone** [reply]
Hmm, so the use-case would be that you upload a ZIP with the media picker and you want to e.g. automatically extract it to a folder still in the Media Library? Then yes, you're on the right track with `IMediaCreatingEventHandler`.

**Reply by MarcBruins** [reply]
Yes, that is the use-case, however the need for a explicit refresh makes this a bit cumbersome for the user. I'm now thinking i should write a custom admin module that lets you upload a ZIP and extract that to the media library. This would be an easier process to understand for the editor than to refresh explicitly.

**Reply by Piedone** [reply]
I see now. Then yes, a custom UI would be better. E.g. you can create your own field for this. Behind the scenes, it can still interact with `IMediaStore` and the other services to store everything in the Media Library.

**Comment by deanmarcussen** [comment, marked as answer]
Yes, the `IMediaCreatingEventHandler` maybe the place to start.

However that handler expects to mutate the incoming stream, possibly to a different format, or size, and return a new stream and filename. 

So it will be a little trickier, as from what you say, you want to extract the contents of the zip archive, and store those contents in a folder.

It should be possible, but you will still have to return a stream from the handler as well, and your handler will probably have to internally resolve the filestore in order to create more entries / folders from the zip archive.

In terms of what you will see in the media library, it will be whatever files you end up creating.

But if you have created multiple files from one archive, and / or extra folders, they won't be visible until you refresh the library.

I would suggest you may find it easier to writing a small admin controller that injected the `IMediaFileStore`, recieved the file, converted it how it needs to be converted, and stored the results in the appropriate location in the media library. 

A custom recipe step might also be an option. It depends how many of these files there are, and how they are produced.

**Reply by MarcBruins** [reply]
This makes the `IMediaCreatingEventHandler` a no go. It needs a refresh of the page which is confusing to the user. I'd rather make an extra `menu item` in which they can upload the zip and then use the `IMediaFileStore` to upload it in the media library.

Can it be only a controller that i add to the project? I followed the steps [add admin menu](https://docs.orchardcore.net/en/dev/docs/guides/add-admin-menu/) and that comes up with a complete new module(which is also fine).

What i don't understand is that when i open the page inside my admin menu it renders the `childOne` perfectly, however i also want the admin menu to persist while the users uploads the content. Any idea on how to do that?

**Comment by MarcBruins** [comment]
I'll close this one and post other questions i have in different discussions. Thanks for the help, this definitely set met on the right track.
